src/main/scala/org/apache/hadoop/fs/sftpwithseek/SFTPConnectionPool.scala

```diff
@@ -0,0 +1,229 @@
+ */
+/** Concurrent/Multiple Connections. */
+object SFTPConnectionPool {
+
+  val LOG: Log = LogFactory.getLog(classOf[SFTPFileSystem])
+
+  /**
+   * Class to capture the minimal set of information that distinguish
+   * between different connections.
+   */
+  private[sftpwithseek] class ConnectionInfo private[sftpwithseek] (
+      val hst: String,
+      var port: Int,
+      val usr: String
+  ) {
+
+    private var host = ""
+    private var user = ""
+    this.host = hst
+    this.user = usr
+    def getHost: String            = host
+    def setHost(hst: String): Unit = this.host = hst
+    def getPort: Int               = port
+    def setPort(prt: Int): Unit    = this.port = prt
+    def getUser: String            = user
+    def setUser(usr: String): Unit = this.user = usr
+
+    def canEqual(a: Any) = a.isInstanceOf[ConnectionInfo]
+
+    override def equals(obj: Any): Boolean = {
+
+      obj match {
+        case obj: ConnectionInfo => {
+          obj.canEqual(this) &&
+          (!this.host.isEmpty && this.host.equalsIgnoreCase(obj.host)) &&
+          (!this.user.isEmpty && this.user.equalsIgnoreCase(obj.user)) &&
+          (this.port >= 0 && this.port == obj.port)
+        }
+        case _ => false
+      }
+
+//      if (this eq obj) return true
+//      if (obj.isInstanceOf[ConnectionInfo])  {
+//        val con = obj.asInstanceOf[ConnectionInfo]
+//        var ret = true
+//        if (this.host == null || !this.host.equalsIgnoreCase(con.host)) ret = false
+//        if (this.port >= 0 && this.port != con.port) ret = false
+//        if (this.user == null || !this.user.equalsIgnoreCase(con.user)) ret = false
+//        ret
+//      }
+//      else false
+    }
+    override def hashCode: Int = {
+      var hashCode = 0
+      if (host != null) hashCode += host.hashCode
+      hashCode += port
+      if (user != null) hashCode += user.hashCode
+      hashCode
+    }
+  }
+}
+class SFTPConnectionPool private[sftpwithseek] ( // Maximum number of allowed live connections. This doesn't mean we cannot
+    // have more live connections. It means that when we have more
+    // live connections than this threshold, any unused connection will be
+    // closed.
+    var maxConnection: Int
+) {
+
+  import SFTPConnectionPool._
+
+  private var liveConnectionCount = 0
+  private var idleConnections     = new util.HashMap[ConnectionInfo, util.HashSet[ChannelSftp]]
+  private var con2infoMap         = new util.HashMap[ChannelSftp, ConnectionInfo]
+  @throws[IOException]
+  private[sftpwithseek] def getFromPool(info: ConnectionInfo): ChannelSftp = {
+    val cons                 = idleConnections.get(info)
+    var channel: ChannelSftp = null
+    if (cons != null && cons.size > 0) {
+      val it = cons.iterator
+      if (it.hasNext) {
+        channel = it.next
+        idleConnections.remove(info)
+        return channel
+      } else throw new IOException("Connection pool error.")
+    }
+    null
```

Comment:
:question:

Reply:
Same as above.

Reply:
Moved the classes to their original Java implemention in: 25d34a0e31b595893e2f9bd3fe392ba9591ef249